File: benchmark/src/main/scala/org/finos/vuu/benchmark/SortBenchmark.scala
package org.finos.vuu.benchmark

import org.finos.toolbox.collection.array.ImmutableArray
import org.finos.toolbox.time.{Clock, DefaultClock}
import org.finos.vuu.core.sort.{GenericSort, GenericSort2}
import org.finos.vuu.core.table.{SimpleDataTable, ViewPortColumnCreator}
import org.finos.vuu.net.{SortDef, SortSpec}
import org.openjdk.jmh.annotations._
import org.openjdk.jmh.runner.Runner
import org.openjdk.jmh.runner.options.OptionsBuilder

import java.io.IOException
import java.util.concurrent.TimeUnit

@State(Scope.Benchmark)
class SortBenchmark {

  import SortBenchmarkHelper._

  var table: SimpleDataTable = null

  @Setup(Level.Invocation)
  def setup(): Unit = {
    table = createBigTable(2_000_000)
  }

  def doSort(table: SimpleDataTable, sort: GenericSort2): ImmutableArray[String] = {
    val viewPortColumns = ViewPortColumnCreator.create(table, table.columns().filter(_.name.equals("exchange")).map(_.name).toList)
    sort.doSort(table, table.primaryKeys, viewPortColumns)
  }

  @Benchmark
  @OutputTimeUnit(TimeUnit.MILLISECONDS)
  @Warmup(iterations = 5)
  @Measurement(iterations = 10)
  @BenchmarkMode(Array(Mode.AverageTime))
  @throws[IOException]
  def sortLargeTable(): Unit = {
    implicit val clock: Clock = new DefaultClock
    val sort = GenericSort2(SortSpec(List(SortDef("exchange", 'A'))), table.getTableDef.columns.filter(_.name == "exchange").toList)
    doSort(table, sort)
  }

  def main(args: Array[String]): Unit = {
    val opts = new OptionsBuilder()
      .include(classOf[SortBenchmark].getSimpleName)
      .warmupIterations(5)
      .measurementIterations(5)
      //.forks(1)
      .build
    new Runner(opts).run
  }

}

object SortRun {
  def main(args: Array[String]): Unit = {
    val benchmark = new SortBenchmark()
    benchmark.setup()
    benchmark.sortLargeTable()
  }
}

File: benchmark/src/main/java/org/finos/vuu/benchmark/SortBenchmark.java
package org.finos.vuu.benchmark;

import org.openjdk.jmh.annotations.*;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

@State(Scope.Benchmark)
public class SortBenchmark {

  public static void main(String[] args) throws Exception {
    final SortBenchmark bm = new SortBenchmark();
    bm.tableSize = 10000;
    bm.setup();
    bm.sortLargeTable();
  }

  private org.finos.vuu.benchmark.sort.SortBenchmark benchmark;

  @Param({ "10000", "100000", "500000", "1000000", "2000000", "5000000" })
  public int tableSize;

  @Setup
  public void setup(){
    benchmark = new org.finos.vuu.benchmark.sort.SortBenchmark();
    System.out.println("Setup: tableSize=" + tableSize);
    benchmark.setup(tableSize);
  }

    @Benchmark
    @OutputTimeUnit(TimeUnit.MILLISECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(1)
    @BenchmarkMode(Mode.SampleTime)
    public void sortLargeTable() throws IOException {
      //System.out.println("Sort: tableSize=" + tableSize);
      benchmark.sortLargeTable();
    }
}


